packages/client/hmi-client/src/services/model-templating.ts

```diff
@@ -79,16 +77,223 @@ export function junctionCleanUp(modelTemplates: ModelTemplates) {
 	modelTemplates.junctions = modelTemplates.junctions.filter(({ edges }) => edges.length > 1);
 }
 
+// Designed to work with the decomposed model templates
+function appendNumberToModelVariables(model: any, number: number) {
+	const newModel = cloneDeep(model);
+	const { states, transitions } = newModel.model;
+	const { rates, initials, parameters, observables } = newModel.semantics.ode;
+
+	newModel.header.name += number;
+	newModel.metadata.templateCard.name += number;
+
+	states.forEach((state) => {
+		state.id += number;
+		state.name += number;
+	});
+
+	transitions.forEach((transition) => {
+		transition.id += number;
+		transition.properties.name += number;
+		transition.input = transition.input.map((input) => input + number);
+		transition.output = transition.output.map((output) => output + number);
+	});
+
+	initials.forEach((initial) => {
+		initial.target += number;
+	});
+
+	parameters.forEach((parameter) => {
+		parameter.id += number;
+	});
+
+	rates.forEach((rate) => {
+		rate.target += number;
+		// Within the expression attributes update the targets to match the new ones
+		rate.expression = rate.expression.replace(/([a-zA-Z])/g, (match) => match + number);
+		rate.expression_mathml = rate.expression_mathml.replace(
+			/<ci>([a-zA-Z])<\/ci>/g,
+			(_, letter: string) => `<ci>${letter + number}</ci>`
+		);
+	});
+
+	observables.forEach((observable) => {
+		observable.id += number;
+		observable.name += number;
+	});
+
+	return newModel;
+}
+
+export function addCard(
+	modelTemplates: ModelTemplates,
+	kernelManager: KernelSessionManager,
+	outputCode: Function,
+	modelTemplate: any
+) {
+	const { name } = modelTemplate.header;
+
+	// If a decomposed card is added, add it to the kernel
+	if (Object.values(DecomposedModelTemplateTypes).includes(name)) {
+		const addTemplateArguments: AddTemplateArguments = {};
+
+		// Append a number to the model variables to avoid conflicts
+		modelTemplate = appendNumberToModelVariables(modelTemplate, modelTemplates.models.length);
```

Comment:
I don't think this will always work if you are basing it on `modelTemplates.models.length`, because through the additional and removal of template actions the length-value will get repeated. 

Having said that, we can put a giant FIXME on this and address it subsequently. As it also partially plays into whether we want to run immediately or bach stuff up in the notebook.